File: my_app/api/seo.py
from fastapi import APIRouter, HTTPException
from ..services.seo_service import (
    analyze_product_url,
    SEOService,
    generate_seo_improvement_suggestions
)
from ..schemas.seo_schema import ProductUrlAnalysisRequest, SEOAnalysisResponse


router = APIRouter(prefix="/api/v1/seo", tags=["SEO"])



@router.post("/analyze-url", response_model=SEOAnalysisResponse)
async def analyze_url_endpoint(request: ProductUrlAnalysisRequest):
    """
    Analyzes a product URL for SEO metrics.
    """
    try:
        result = await analyze_product_url(
            url=str(request.url), primary_keyword=request.primary_keyword
        )
        if result.get("error"):
            raise HTTPException(status_code=400, detail=result.get("error"))
        return SEOAnalysisResponse(**result)
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"An unexpected error occurred: {str(e)}"
        )

# Individual endpoints for each feature
from fastapi import Body

@router.post("/title-analysis")
async def title_analysis(request: ProductUrlAnalysisRequest):
    result = await analyze_product_url(url=str(request.url), primary_keyword=request.primary_keyword)
    return {"title": result.get("title")}

@router.post("/meta-description-analysis")
async def meta_description_analysis(request: ProductUrlAnalysisRequest):
    result = await analyze_product_url(url=str(request.url), primary_keyword=request.primary_keyword)
    return {"meta_description": result.get("meta_description")}

@router.post("/readability-score")
async def readability_score(request: ProductUrlAnalysisRequest):
    result = await analyze_product_url(url=str(request.url), primary_keyword=request.primary_keyword)
    return {"readability_score": result.get("readability_score")}

@router.post("/keyword-analysis")
async def keyword_analysis(request: ProductUrlAnalysisRequest):
    result = await analyze_product_url(url=str(request.url), primary_keyword=request.primary_keyword)
    return {
        "lsi_keywords": result.get("lsi_keywords"),
        "tfidf_keywords": result.get("tfidf_keywords")
    }

@router.post("/heading-analysis")
async def heading_analysis(request: ProductUrlAnalysisRequest):
    # For demo, extract main text from URL and analyze headings
    result = await analyze_product_url(url=str(request.url), primary_keyword=request.primary_keyword)
    # You may want to pass the HTML or main_text to analyze_headings
    return {"headings": SEOService.analyze_headings(result.get("meta_description", ""))}

@router.post("/image-analysis")
async def image_analysis(request: ProductUrlAnalysisRequest):
    result = await analyze_product_url(url=str(request.url), primary_keyword=request.primary_keyword)
    return {"images": SEOService.analyze_images(result.get("meta_description", ""))}

@router.post("/link-analysis")
async def link_analysis(request: ProductUrlAnalysisRequest, your_domain: str = Body(...)):
    result = await analyze_product_url(url=str(request.url), primary_keyword=request.primary_keyword)
    return {"links": SEOService.analyze_links(result.get("meta_description", ""), your_domain)}

@router.post("/ai-suggestions")
async def ai_suggestions(request: ProductUrlAnalysisRequest):
    result = await analyze_product_url(url=str(request.url), primary_keyword=request.primary_keyword)
    # You may need to pass a valid OpenAI client here
    suggestions = generate_seo_improvement_suggestions(openai_client=None, analysis_results=result)
    return {"suggestions": suggestions}

File: frontend/src/api/seo.ts
import axios from 'axios';

export interface SEOAnalysisRequest {
  url: string;
  primary_keyword: string;
}

export interface SEOAnalysisResponse {
  url: string;
  title: string;
  meta_description?: string;
  readability_score?: number;
  lsi_keywords: any[];
  tfidf_keywords: any[];
  overall_score?: number;
}

export async function analyzeProductUrl(request: SEOAnalysisRequest): Promise<SEOAnalysisResponse> {
  const response = await axios.post<SEOAnalysisResponse>(
    '/api/v1/seo/analyze-url',
    request
  );
  return response.data;
}

export async function getTitleAnalysis(request: SEOAnalysisRequest): Promise<{ title: string }> {
  const response = await axios.post<{ title: string }>(
    '/api/v1/seo/title-analysis',
    request
  );
  return response.data;
}

export async function getMetaDescriptionAnalysis(request: SEOAnalysisRequest): Promise<{ meta_description: string }> {
  const response = await axios.post<{ meta_description: string }>(
    '/api/v1/seo/meta-description-analysis',
    request
  );
  return response.data;
}

export async function getReadabilityScore(request: SEOAnalysisRequest): Promise<{ readability_score: number }> {
  const response = await axios.post<{ readability_score: number }>(
    '/api/v1/seo/readability-score',
    request
  );
  return response.data;
}

export async function getKeywordAnalysis(request: SEOAnalysisRequest): Promise<{ lsi_keywords: any[]; tfidf_keywords: any[] }> {
  const response = await axios.post<{ lsi_keywords: any[]; tfidf_keywords: any[] }>(
    '/api/v1/seo/keyword-analysis',
    request
  );
  return response.data;
}

export async function getHeadingAnalysis(request: SEOAnalysisRequest): Promise<{ headings: any }> {
  const response = await axios.post<{ headings: any }>(
    '/api/v1/seo/heading-analysis',
    request
  );
  return response.data;
}

export async function getImageAnalysis(request: SEOAnalysisRequest): Promise<{ images: any }> {
  const response = await axios.post<{ images: any }>(
    '/api/v1/seo/image-analysis',
    request
  );
  return response.data;
}

export async function getLinkAnalysis(request: SEOAnalysisRequest, your_domain: string): Promise<{ links: any }> {
  const response = await axios.post<{ links: any }>(
    '/api/v1/seo/link-analysis',
    request,
    { params: { your_domain } }
  );
  return response.data;
}

export async function getAISuggestions(request: SEOAnalysisRequest): Promise<{ suggestions: string }> {
  const response = await axios.post<{ suggestions: string }>(
    '/api/v1/seo/ai-suggestions',
    request
  );
  return response.data;
}
